spot-vaults/contracts/_strategies/SpotAppraiser.sol

```diff
@@ -0,0 +1,257 @@
+pragma solidity ^0.8.24;
+
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+import { Math } from "@openzeppelin/contracts/utils/math/Math.sol";
+
+import { IERC20Metadata } from "@openzeppelin/contracts/token/ERC20/extensions/IERC20Metadata.sol";
+import { ITranche } from "@ampleforthorg/spot-contracts/contracts/_interfaces/buttonwood/ITranche.sol";
+import { IBondController } from "@ampleforthorg/spot-contracts/contracts/_interfaces/buttonwood/IBondController.sol";
+import { IPerpetualTranche } from "@ampleforthorg/spot-contracts/contracts/_interfaces/IPerpetualTranche.sol";
+import { IBalancer } from "@ampleforthorg/spot-contracts/contracts/_interfaces/IBalancer.sol";
+import { IChainlinkOracle } from "../_interfaces/external/IChainlinkOracle.sol";
+import { IAmpleforth } from "../_interfaces/external/IAmpleforth.sol";
+import { IAmpleforthOracle } from "../_interfaces/external/IAmpleforthOracle.sol";
+import { IAMPL } from "../_interfaces/external/IAMPL.sol";
+import { IBillBrokerPricingStrategy } from "../_interfaces/IBillBrokerPricingStrategy.sol";
+import { Range } from "../_interfaces/BillBrokerTypes.sol";
+import { InvalidPerc, InvalidSeniorCDRBound } from "../_interfaces/BillBrokerErrors.sol";
+
+/**
+ * @title SpotAppraiser
+ *
+ * @notice Pricing strategy adapter for a BillBroker vault which accepts
+ *         SPOT (as the perp token) and dollar tokens like USDC.
+ *
+ *         AMPL is the underlying token for SPOT.
+ *         The market price of AMPL is mean reverting and eventually converges to its target.
+ *         However, it can significantly deviate from the target in the near term.
+ *
+ *         SPOT is a perpetual claim on AMPL senior tranches. Insofar as SPOT is fully backed by
+ *         healthy senior tranches, we can price spot reliably using the following strategy:
+ *
+ *         SPOT_PRICE = MULTIPLIER * AMPL_TARGET
+ *         MULTIPLIER = spot.getTVL() / spot.totalSupply(), which is it's enrichment/debasement factor.
+ *         To know more, read the spot documentation.
+ *
+ *         We get the AMPL target price from Ampleforth's CPI oracle,
+ *         which is also used by the protocol to adjust AMPL supply through rebasing.
+ *
+ *         And the MULTIPLIER is directly queried from the SPOT contract.
+ *
+ */
+contract SpotAppraiser is Ownable, IBillBrokerPricingStrategy {
+    //-------------------------------------------------------------------------
+    // Libraries
+    using Math for uint256;
+
+    //-------------------------------------------------------------------------
+    // Constants & Immutables
+
+    uint256 private constant DECIMALS = 18;
+    uint256 private constant ONE = (10 ** DECIMALS);
+    uint256 public constant CL_ORACLE_DECIMALS = 8;
+    uint256 public constant CL_ORACLE_STALENESS_TRESHOLD_SEC = 3600 * 24; // 1 day
+    uint256 public constant USD_LOWER_BOUND = (99 * ONE) / 100; // 0.99$
+    uint256 public constant AMPL_DUST_AMT = 1000; // 1000 AMPL
```

Comment:
Generally speaking, it's good style to not have internal representations based on decimals.

If I were developing with this code, I'd instinctually compare AMPL_DUST_AMT to an AMPL balance, but I can't do that as it is. Decimals are really meant for display formatting in UIs.

Reply:
Makes sense, we can use AMPL balance here